src/array/cpu/spmat_op_impl.cc

```diff
@@ -0,0 +1,635 @@
+  std::vector<bool> filter_;
+  // The hashmap from old vid to new vid
+  std::unordered_map<IdType, IdType> oldv2newv_;
+};
+
+struct PairHash {
+  template <class T1, class T2>
+  std::size_t operator() (const std::pair<T1, T2>& pair) const {
+    return std::hash<T1>()(pair.first) ^ std::hash<T2>()(pair.second);
+  }
+};
+
+template <typename DType>
+inline runtime::NDArray VecToNDArray(const std::vector<DType>& vec,
+    DLDataType dtype, DLContext ctx) {
+  const int64_t len = vec.size();
+  NDArray ret_arr = NDArray::Empty({len}, dtype, ctx);
+  DType* ptr = static_cast<DType*>(ret_arr->data);
+  std::copy(vec.begin(), vec.end(), ptr);
+  return ret_arr;
+}
+
+inline bool CSRHasData(CSRMatrix csr) {
+  return csr.data.defined();
+}
+
+inline bool COOHasData(COOMatrix csr) {
+  return csr.data.defined();
+}
+}  // namespace
+
+///////////////////////////// CSRIsNonZero /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+bool CSRIsNonZero(CSRMatrix csr, int64_t row, int64_t col) {
+  CHECK(row >= 0 && row < csr.num_rows) << "Invalid row index: " << row;
+  CHECK(col >= 0 && col < csr.num_cols) << "Invalid col index: " << col;
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices_data = static_cast<IdType*>(csr.indices->data);
+  for (IdType i = indptr_data[row]; i < indptr_data[row + 1]; ++i) {
+    if (indices_data[i] == col) {
+      return true;
+    }
+  }
+  return false;
+}
+
+template bool CSRIsNonZero<kDLCPU, int32_t>(CSRMatrix, int64_t, int64_t);
+template bool CSRIsNonZero<kDLCPU, int64_t>(CSRMatrix, int64_t, int64_t);
+
+template <DLDeviceType XPU, typename IdType>
+NDArray CSRIsNonZero(CSRMatrix csr, NDArray row, NDArray col) {
+  const auto rowlen = row->shape[0];
+  const auto collen = col->shape[0];
+  const auto rstlen = std::max(rowlen, collen);
+  NDArray rst = NDArray::Empty({rstlen}, row->dtype, row->ctx);
+  IdType* rst_data = static_cast<IdType*>(rst->data);
+  const IdType* row_data = static_cast<IdType*>(row->data);
+  const IdType* col_data = static_cast<IdType*>(col->data);
+  const int64_t row_stride = (rowlen == 1 && collen != 1) ? 0 : 1;
+  const int64_t col_stride = (collen == 1 && rowlen != 1) ? 0 : 1;
+  for (int64_t i = 0, j = 0; i < rowlen && j < collen; i += row_stride, j += col_stride) {
+    *(rst_data++) = CSRIsNonZero<XPU, IdType>(csr, row_data[i], col_data[j])? 1 : 0;
+  }
+  return rst;
+}
+
+template NDArray CSRIsNonZero<kDLCPU, int32_t>(CSRMatrix, NDArray, NDArray);
+template NDArray CSRIsNonZero<kDLCPU, int64_t>(CSRMatrix, NDArray, NDArray);
+
+///////////////////////////// CSRHasDuplicate /////////////////////////////
+
+template <DLDeviceType XPU, typename IdType>
+bool CSRHasDuplicate(CSRMatrix csr) {
+  const IdType* indptr_data = static_cast<IdType*>(csr.indptr->data);
+  const IdType* indices_data = static_cast<IdType*>(csr.indices->data);
+  for (IdType src = 0; src < csr.num_rows; ++src) {
```

Comment:
theoretically, the number of rows can still go above 2B nodes and the 32-bit CSR matrix still works fine.